### PR TITLE
add internal opCode implementation

### PR DESCRIPTION
**Pull request by user-dob**
- add Memory
- add gas logic
- add use memory logic
- add ethereum testdata

- https://trello.com/c/pG1XSLUs/8-crypto
- https://trello.com/c/w9wPkmXo/20-interpreter
- https://trello.com/c/YDs9WxYE/11-storage-and-execution

### CODE REVIEW

File: src/Interpreter.ts [hunk 1]
```diff
@@ -20,12 +28,61 @@ export class Interpreter {
...
+        }
+
+        this.state.programCounter++;
+
+        opInst.isAsync
+            ? await opInst.opFn(this.state)
+            : opInst.opFn(this.state);
+    }
+
+    private getOpInst(opCode: OpCode): OpInst {
```
Comment (macaqueee) [inline src/Interpreter.ts]: how about to rename to something more readable.

File: src/Memory.ts [hunk 2]
```diff
@@ -1,4 +1,78 @@
 
+/**
+ * Memory implements a simple memory model
+ * for the ethereum virtual machine.
+ */
 export class Memory {
 
+    // TODO replace to Buffer
+    _store: number[];
```
Comment (macaqueee) [inline src/Memory.ts]: maybe use generic and multiple implementations of Memory? (for Array and for Buffer)

File: src/interpreter/Interpreter.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import { EEI } from 'src/EEI';
+import { ERROR, VmError, VmStop } from 'src/interpreter/exceptions';
+import { VMResult } from 'src/VM';
+
+export class Interpreter {
+
+    private readonly state: State;
+
+    constructor(context: IContext, storage: IStorage) {
+        const eei = new EEI(context);
```
Comment (Verbalist) [inline src/interpreter/Interpreter.ts:16]: EEI rename class to full name

File: src/interpreter/Interpreter.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+    private getValidJumpDests(code: Buffer): Set<number> {
+        const jumps = new Set<number>();
+
+        for (let i = 0; i < code.length; i++) {
+            const instruction = this.getInstruction(code[i]);
+
+            // no destinations into the middle of PUSH
+            if (isPush(instruction.opCode)) {
+                i += code[i] - 0x5f;
```
Comment (Verbalist) [inline src/interpreter/Interpreter.ts]: MAGIC NUMBER exclude to const

Comment (Verbalist) [reply src/interpreter/Interpreter.ts]: or wrap to util

**Review by Verbalist — APPROVED** [review]

**Review by macaqueee — APPROVED** [review]

**Pull request merged**